utils/script-luajit.c

```diff
@@ -0,0 +1,425 @@
+static void (*dllua_remove)(lua_State *L, int index);
+
+static void (*dllua_getfield)(lua_State *L, int index, const char *k);
+static int (*dllua_type)(lua_State *L, int index);
+static void (*dllua_settop)(lua_State *L, int index);
+
+void __attribute__((weak)) lua_getfield(lua_State *L, int index, const char *k)
+{
+	dllua_getfield(L, index, k);
+}
+
+int __attribute__((weak)) lua_type(lua_State *L, int index)
+{
+	return dllua_type(L, index);
+}
+
+void __attribute__((weak)) lua_createtable(lua_State *L, int narr, int nrec)
+{
+	dllua_createtable(L, narr, nrec);
+}
+
+const __attribute__((weak)) char *lua_tolstring(lua_State *L, int index, size_t *len)
+{
+	return dllua_tolstring(L, index, len);
+}
+
+void __attribute__((weak)) lua_settop(lua_State *L, int index)
+{
+	return dllua_settop(L, index);
+}
+
+static void setup_common_context(struct script_context *sc_ctx)
+{
+	lua_newtable(L); // * ctx
+	dllua_pushstring(L, "tid"); // * ctx 'tid'
+	dllua_pushinteger(L, sc_ctx->tid); // * ctx 'tid' tid
+	dllua_settable(L, -3); // * ctx
+	dllua_pushstring(L, "depth"); // * ctx 'depth'
+	dllua_pushinteger(L, sc_ctx->depth); // * ctx 'depth' depth
+	dllua_settable(L, -3); // * ctx
+	dllua_pushstring(L, "timestamp"); // * ctx 'timestamp'
+	dllua_pushinteger(L, sc_ctx->timestamp); // * ctx 'timestamp' timestamp
+	dllua_settable(L, -3); // * ctx
+	dllua_pushstring(L, "duration"); // * ctx 'duration'
+	dllua_pushinteger(L, sc_ctx->duration); // * ctx 'duration' duration
+	dllua_settable(L, -3); // * ctx
+	dllua_pushstring(L, "address"); // * ctx 'address'
+	dllua_pushinteger(L, sc_ctx->address); // * ctx 'address' address
+	dllua_settable(L, -3); // * ctx
+	dllua_pushstring(L, "name"); // * ctx 'name'
+	dllua_pushstring(L, sc_ctx->name); // * ctx 'name' name
+	dllua_settable(L, -3); // * ctx
+}
+
+static void setup_argument_context(bool is_retval, struct script_context *sc_ctx)
+{
+	struct uftrace_arg_spec *spec;
+	void *data = sc_ctx->argbuf;
+	union script_arg_val val;
+	int count = 0;
+
+	list_for_each_entry(spec, sc_ctx->argspec, list) {
+		/* skip unwanted arguments or retval */
+		if (is_retval != (spec->idx == RETVAL_IDX))
+			continue;
+
+		count++;
+	}
+	if (count == 0)
+		return;
+
+	if (is_retval)
+		dllua_pushstring(L, "retval"); // * ctx 'retval'
+	else
+		dllua_pushstring(L, "args"); // * ctx 'args'
+	lua_newtable(L); // * ctx '<retval-or-args>' <retval-or-args-table>
+
+	count = 0;
+	list_for_each_entry(spec, sc_ctx->argspec, list) {
+		const int null_str = -1;
+		unsigned short slen;
+		char ch_str[2];
+		char *str;
+		double dval;
+
+		/* skip unwanted arguments or retval */
+		if (is_retval != (spec->idx == RETVAL_IDX))
+			continue;
+
+		/* reset the value */
+		memset(val.v, 0, sizeof(val));
+
+		switch (spec->fmt) {
+		case ARG_FMT_AUTO:
+		case ARG_FMT_SINT:
+		case ARG_FMT_UINT:
+		case ARG_FMT_HEX:
+		case ARG_FMT_PTR:
+		case ARG_FMT_ENUM:
+			memcpy(val.v, data, spec->size);
+			switch (spec->size) {
```

Comment:
I think you just can use val.v without the switch statement.

Reply:
Thoughts?

Reply:
memcpy to val.v only occurs in the cases of ARG_FMT_FLOAT and ARG_FMT_ENUM

Reply:
As val.v is zero initialized and copied only meaningful size, you can just use `dllua_pushinteger(L, val.v)` regardless of spec->size IMO.

Reply:
Hmm I'm kind of worried about endian-safeness. Would it be safe in both big-endian and little-endian systems?

Reply:
I'm not sure.  OK let's just leave it.